curp/src/server/bg_tasks.rs

```diff
@@ -110,57 +89,29 @@ async fn bg_tick<C: Command + 'static, Conn: ConnectInterface, ExeTx: CmdExeSend
 
     let mut ticker = tokio::time::interval(heartbeat_interval);
     ticker.set_missed_tick_behavior(MissedTickBehavior::Skip);
+    #[allow(clippy::integer_arithmetic, unused_attributes)] // tokio internal triggered
     loop {
         let _now = ticker.tick().await;
-        let task = {
-            let state_c = Arc::clone(&state);
-            let state_r = state.upgradable_read();
-            if state_r.is_leader() {
-                if state_r
-                    .hb_opt
-                    .compare_exchange(true, false, Ordering::Acquire, Ordering::Relaxed)
-                    .is_err()
-                {
-                    let resps = bcast_heartbeats(connects.clone(), state_r, rpc_timeout);
-                    Either::Left(handle_heartbeat_responses(
-                        resps,
-                        state_c,
-                        Arc::clone(&timeout),
-                    ))
-                } else {
-                    continue;
-                }
-            } else {
-                let tick = state_r.election_tick.fetch_add(1, Ordering::AcqRel);
-                if tick
-                    < (if state_r.role() == ServerRole::Follower {
-                        follower_timeout_ticks
-                    } else {
-                        candidate_timeout_ticks
-                    })
-                {
-                    continue;
-                }
-                // re-generate timeout to avoid split vote forever
-                follower_timeout_ticks = thread_rng()
-                    .gen_range(follower_timeout_ticks_base..(2 * follower_timeout_ticks_base));
-                candidate_timeout_ticks = thread_rng()
-                    .gen_range(candidate_timeout_ticks_base..(2 * candidate_timeout_ticks_base));
-                state_r.reset_election_tick(); // reset tick
-
-                let resps = bcast_votes(connects.clone(), state_r, rpc_timeout);
-                Either::Right(handle_vote_responses(resps, state_c))
+        let action = curp.tick();
+        match action {
+            TickAction::Heartbeat(hbs) => {
+                let resps = bcast_heartbeats(&connects, hbs, rpc_timeout);
+                handle_heartbeat_responses(curp.as_ref(), resps, &calibrate_tx).await;
             }
-        };
-        task.await;
+            TickAction::Votes(votes) => {
+                let resps = bcast_votes(&connects, votes, rpc_timeout);
+                handle_vote_responses(curp.as_ref(), resps).await;
+            }
+            TickAction::Nothing => {}
+        }
     }
 }
 
 /// Fetch commands need to be synced and add them to the log
-async fn bg_get_sync_cmds<C: Command + 'static, ExeTx: CmdExeSenderInterface<C>>(
-    state: StateRef<C, ExeTx>,
+async fn bg_get_sync_cmds<C: Command + 'static>(
+    curp: Arc<RawCurp<C>>,
     sync_rx: flume::Receiver<SyncMessage<C>>,
-    ae_tx: mpsc::UnboundedSender<usize>,
+    ae_trigger: mpsc::UnboundedSender<usize>,
```

Comment:
We can use `ae_tx` to replace `ae_trigger`.

Reply:
Fixed.